src/attesters/sgx-ecdsa/collect_endorsements.c

```diff
@@ -0,0 +1,120 @@
+		ret = ENCLAVE_ATTESTER_ERR_INVALID;
+		goto err;
+	}
+	/* fence after boundary check */
+	sgx_lfence();
+
+	/* Copy collateral struct from untrusted-app to enclave memory */
+	sgx_ql_qve_collateral_t collateral;
+	memcpy(&collateral, collateral_untrusted, sizeof(sgx_ql_qve_collateral_t));
+
+	/* Copy fields from collateral to endorsements->ecdsa */
+	sgx_ql_qve_collateral_t *c = &collateral;
+	sgx_ecdsa_attestation_collateral_t *e = &endorsements->ecdsa;
+
+	e->version = c->version;
+	#define COPY_ENDORSEMENT_FIELD(value_field, size_field)                       \
+		{                                                                     \
+			/* Memory boundary check */                                   \
+			if (!sgx_is_outside_enclave(c->value_field, c->size_field)) { \
+				ret = ENCLAVE_ATTESTER_ERR_INVALID;                   \
+				goto err;                                             \
+			}                                                             \
+			sgx_lfence();                                                 \
+			e->value_field = malloc(c->size_field);                       \
+			if (!e->value_field) {                                        \
+				ret = ENCLAVE_ATTESTER_ERR_NO_MEM;                    \
+				goto err;                                             \
+			}                                                             \
+			memcpy(e->value_field, c->value_field, c->size_field);        \
+			e->size_field = c->size_field;                                \
+		}
+
+	COPY_ENDORSEMENT_FIELD(pck_crl_issuer_chain, pck_crl_issuer_chain_size);
+
+	COPY_ENDORSEMENT_FIELD(root_ca_crl, root_ca_crl_size);
+
+	COPY_ENDORSEMENT_FIELD(pck_crl, pck_crl_size);
+
+	COPY_ENDORSEMENT_FIELD(tcb_info_issuer_chain, tcb_info_issuer_chain_size);
+
+	COPY_ENDORSEMENT_FIELD(tcb_info, tcb_info_size);
+
+	COPY_ENDORSEMENT_FIELD(qe_identity_issuer_chain, qe_identity_issuer_chain_size);
+
+	COPY_ENDORSEMENT_FIELD(qe_identity, qe_identity_size);
+
+	RTLS_DEBUG(
+		"version: %d, pck_crl_issuer_chain_size: %d, root_ca_crl_size: %d, pck_crl_size: %d, tcb_info_issuer_chain_size: %d, tcb_info_size: %d, qe_identity_issuer_chain_size: %d, qe_identity_size: %d\n",
+		c->version, c->pck_crl_issuer_chain_size, c->root_ca_crl_size, c->pck_crl_size,
+		c->tcb_info_issuer_chain_size, c->tcb_info_size, c->qe_identity_issuer_chain_size,
+		c->qe_identity_size);
+
+	ret = ENCLAVE_ATTESTER_ERR_NONE;
+err:
+	if (collateral_untrusted) {
+		enclave_attester_err_t q_ret = sgx_status =
+			ocall_tee_qv_free_collateral(&q_ret, collateral_untrusted);
+		if (SGX_SUCCESS != sgx_status || ENCLAVE_ATTESTER_ERR_NONE != q_ret) {
+			RTLS_ERR("tee_qv_free_collateral(): 0x%04x, 0x%04x\n", sgx_status, q_ret);
+		}
+	}
+
+	if (ret != ENCLAVE_ATTESTER_ERR_NONE)
+		free_endorsements(evidence->type, endorsements);
+	return ret;
+}
+
+#else
+
+enclave_attester_err_t sgx_ecdsa_collect_endorsements(enclave_attester_ctx_t *ctx,
```

Comment:
How to `collect_endorsements` in Occlum Libos Mode?

Reply:
Ok, I've created an issue in occlum project. https://github.com/occlum/occlum/issues/1102